topics/integrate-in-existing-app.md

```diff
@@ -1,242 +1,423 @@
-    }
+    ```kotlin
+    override fun onCreate(savedInstanceState: Bundle?) {
+       super.onCreate(savedInstanceState)
 
-    private var view: View? = null
-    private var lastCommand: View.() -> Unit =
-            { displayLoading(false) }
-    private fun commandView(command: View.() -> Unit) {
-        lastCommand = command
-        view?.command()
+       Log.i("Login Activity", "Hello from shared module: " + (Greeting().greeting()))
+   
     }
+    ```
+   
+5. Search for `Hello` in the log, and you'll find a greeting from the shared module.
 
-    fun attach(view: View) { this.view = view.apply(lastCommand) }
-    fun detach() { this.view = null }
-
-    fun login(username: String, password: String) {
-        MainScope().launch {
-            try {
-                commandView { displayLoading(true) }
-                getNetwork().login(username, password) // suspending
-                commandView { goToNextScreen() }
-            } catch (ex: LoginException) {
-                commandView {
-                    displayLoading(false)
-                    displayError(ex.code)
-                }
-            }
-        }
-    }
-}
-```
+    ![Greeting from the shared module](shared-module-greeting.png)
 
-Note the `commandView` and `lastCommand` mechanism, which allows a view to detach and re-attach, for example for configuration changes on Android.
+### Make the business logic cross platform
 
-#### MVI for declarative UI frameworks {initial-collapse-state="collapsed"}
+You can now extract the business logic code to the KMM shared module and make it platform independent. This is necessary for reusing it for both Android and iOS.
+ 
+1. Move the business logic code `com.jetbrains.simplelogin.androidapp.data` from the `app` directory to the `com.jetbrains.simplelogin.shared` package in the `shared/src/commonMain` directory.
+   You can drag and drop the package or refactor it by moving everything from one directory to another.
+   
+    ![Drag and drop the package with the business logic code](moving-business-logic.png){width=350}
 
-_Model-View-Intent_ (_MVI_) is the natural evolution of MVP when working with declarative UI frameworks (although it also works with legacy UI frameworks). 
-It is therefore recommended to be used with Swift UI or Jetpack Compose.
+2. When Android Studio asks what you'd like to do, select to move the package, and then approve refactoring.
 
-In MVI, the entire UI structure is described in one tree. Here is an example of a simple MVI presenter:
+    ![Refactor the buiness logic package](refactor-business-logic-package.png){width=500}
 
-```kotlin
-class LoginPresenter {
-    sealed class Model {
-        object Form : Model()
-        object Loading : Model()
-        data class Error(val code: Int): Model()
-        object GoToNextScreen : Model()
-    }
-    interface View {
-        fun displayModel(model: Model)
-    }
+3. Ignore all warnings about platform-dependent code and click **Continue**.
```

Comment:
Maybe it's worth explaining why we have these warnings? Like "IDE notifies us that some classes and methods will be unavailable in the shared module, but we will fix this later" or smth like this

Reply:
Ok, I agree about the explanation. My intention was not to explain the warning, but _why_ is we get some. But maybe it's too much and we have pretty comprehensive explanations about platform-specific apis below.

I would rather keep the screenshot, because as a user I'm used to using screenshots to check if I'm following the right directions, like a save points) :D

Reply:
I kept the screenshot but without explanation.